### PR TITLE
Add thread safe rate limiting

### PR DESCRIPTION
**Pull request by claudiosv**

### CODE REVIEW

File: lmwrapper/huggingface_wrapper.py [hunk 1]
```diff
@@ -243,6 +319,7 @@ def new_call(*args, **kwargs):
                 # TODO: some models require an attention mask, others not?
                 # **encoded_input
                 generation_config=gen_config,
+                stopping_criteria=stopping_criteria,
```
Comment (DNGros) [inline lmwrapper/huggingface_wrapper.py:322]: might need to do the "backtracking"/chopping on the output as discussed on Teams. Might take either stopping a token late (so the whole stop prefix appears in the generation output) or detecting if stopped early in order to know if need to chop or not.

This whole thing also just gets messy with logprobs too. I'm a fan of simplified versions with exceptions for now if a quick fix is not appearent.

File: lmwrapper/abstract_predictor.py [hunk 2]
```diff
@@ -17,12 +17,24 @@ def _predict_definately_cached(
...
 class LmPredictor:
+    _rate_limit: Optional[RateLimit] = None
+
     def __init__(
         self,
         cache_default: bool = False,
     ):
         self._cache_default = cache_default
 
+    def configure_global_ratelimit(max_count=1, per=1, greedy=False) -> None:
```
Comment (DNGros) [inline lmwrapper/abstract_predictor.py:28]: Please either document the units of `per` is seconds, or use a descriptive name (`seconds_per_max_count` or some other wording). Also not really sure what `greedy` does.

File: lmwrapper/huggingface_wrapper.py [hunk 3]
```diff
@@ -166,6 +165,57 @@ def get_full_tokens(self):
...
+            assert tokenizer
+        else:
+            assert all(
+                isinstance(stop_sequence, int) for stop_sequence in self.stop_sequences
+            )
+
+    def __call__(
+        self, input_ids: torch.LongTensor, scores: torch.FloatTensor, **kwargs
+    ) -> bool:
+        assert input_ids.shape[0] == 1
```
[the remark this answers is not in the document]
Comment (claudiosv) [reply lmwrapper/huggingface_wrapper.py:202]: Batching is not supported, it should raise an exception in `_predict_maybe_cached`.

File: test/test_huggingface.py [hunk 4]
```diff
@@ -29,6 +29,56 @@ class Models(StrEnum):
...
+    # Let's make sure we get a stop token in our prompt normally
+    prompt = LmPrompt(
+        prompt_str,
+        max_tokens=300,
+        cache=False,
+        temperature=0,
+        stop=["I like to eat candy"],
+    )
+    out = lm.predict(prompt)
+    assert "I like to eat candy" not in out.completion_text
```
Comment (DNGros) [inline test/test_huggingface.py:79]: the test here is a bit more subtle though. We also don't "I like to eat"in there either.

File: test/test_huggingface.py [hunk 5]
```diff
@@ -29,6 +29,115 @@ class Models(StrEnum):
...
+        prompt_str,
+        max_tokens=15,
+        cache=False,
+        temperature=0,
+    )
+    out = lm.predict(prompt)
+    assert "Italy" in out.completion_text
+
+    prompt = LmPrompt(
+        prompt_str, max_tokens=15, cache=False, temperature=0, stop=["Italy"]
```
Comment (DNGros) [inline test/test_huggingface.py:53]: Also this test should be include "Ita" in there two for that point.

File: test/test_huggingface.py (hunk 4 above)
Comment (DNGros) [reply test/test_huggingface.py:79]: (consider just asserting the expected text)

Comment (claudiosv) [reply test/test_huggingface.py:79]: We don't?

**Pull request closed without merging**